### PR TITLE
Fix IPv6 to_string [10017]

### PR DESCRIPTION
**Pull request by jparisu**
Signed-off-by: jparisu <javierparis@eprosima.com>

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3161/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/379/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2821/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3193/)

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
This would have been detected if we had unit tests for IPLocator. Would you mind adding it?

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3202/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/416/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2867/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3234/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3206/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/419/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2874/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3238/)

File: test/unittest/utils/IPLocatorTests.cpp [hunk 1]
```diff
@@ -0,0 +1,867 @@
...
+}
+
+TEST_F(IPLocatorTests, setIPv4)
+{
+    Locator_t locator;
+    Locator_t probe_locator;
+    IPLocator::createLocator(LOCATOR_KIND_UDPv4, ipv4_lo_address, port1, locator);
+
+    // setIPv4 char*
+    // TODO
```
Comment (EduPonz) [inline test/unittest/utils/IPLocatorTests.cpp:342]: Please address all TODOs

File: test/unittest/utils/IPLocatorTests.cpp [hunk 2]
```diff
@@ -0,0 +1,867 @@
...
+{
+    Locator_t locator;
+    ASSERT_TRUE(IPLocator::setLogicalPort(locator, port1));
+    ASSERT_EQ(IPLocator::getLogicalPort(locator), port1);
+    ASSERT_NE(IPLocator::getLogicalPort(locator), port2);
+    ASSERT_TRUE(IPLocator::setLogicalPort(locator, port2));
+    ASSERT_EQ(IPLocator::getLogicalPort(locator), port2);
+}
+
+TEST_F(IPLocatorTests, setPhysicalPort)
```
Comment (EduPonz) [inline test/unittest/utils/IPLocatorTests.cpp:562]:
```suggestion
TEST_F(IPLocatorTests, PhysicalPort)
```

**Review by EduPonz — CHANGES_REQUESTED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3215/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/427/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2879/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3247/)

File: src/cpp/utils/IPLocator.cpp [hunk 3]
```diff
@@ -191,97 +189,154 @@ bool IPLocator::setIPv6(
         Locator_t& locator,
         const std::string& ipv6)
 {
-    //std::string _ipv6 = IPFinder::getIPv6Address(ipv6);
-    std::vector<std::string> hexdigits;
+    if (!IPv6isCorrect(ipv6))
+    {
+        // ERROR
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/utils/IPLocator.cpp:194]: It is an API method and I think it should not change the output.
In last implementation, and IPv4 does not send an Error message.

File: src/cpp/utils/IPLocator.cpp [hunk 4]
```diff
@@ -88,14 +80,18 @@ bool IPLocator::setIPv4(
...
     int a, b, c, d; //to store the 4 ints
     char ch; //to temporarily store the '.'
 
     if (ss >> a >> ch >> b >> ch >> c >> ch >> d)
     {
-        LOCATOR_ADDRESS_INVALID(locator.address);
+        if (a > 255 || b > 255 || c > 255 || d > 255)
+        {
+            return false;
+        }
```
Comment (imontesino) [inline src/cpp/utils/IPLocator.cpp:94]: The memory cleaning should only apply to the  last 4 octets, those corresponding to the IPv4 adress

**Comment by MiguelCompany** [comment]
@imontesino 400 warnings on windows and 3 on Mac

Only lines 307 and 308 on Locator.h are relevant

File: include/fastdds/rtps/common/Locator.h [hunk 5]
```diff
@@ -263,78 +291,78 @@ inline std::istream& operator >>(
 
     if (s)
     {
-        char punctuation;
-        unsigned short kind;
-        unsigned short value;
+        unsigned int bracket_index;
```
Comment (MiguelCompany) [inline include/fastdds/rtps/common/Locator.h:294]:
```suggestion
```

File: include/fastdds/rtps/common/Locator.h [hunk 6]
```diff
@@ -263,78 +291,78 @@ inline std::istream& operator >>(
...
         {
             input.exceptions(excp_mask | std::ios_base::failbit | std::ios_base::badbit);
 
-            input >> kind >> punctuation;
-            loc.kind = kind;
+            // First check the locator kind
+            input >> str_locator;
 
-            if (kind == LOCATOR_KIND_SHM)
+            bracket_index = str_locator.find(']');
```
Comment (MiguelCompany) [inline include/fastdds/rtps/common/Locator.h:307]:
```suggestion
            auto bracket_index = str_locator.find(']');
```

**Review by MiguelCompany — CHANGES_REQUESTED** [review]
This seems to work (at least on windows)

File: include/fastdds/rtps/common/Locator.h [hunk 7]
```diff
@@ -263,78 +291,78 @@ inline std::istream& operator >>(
 
     if (s)
     {
-        char punctuation;
-        unsigned short kind;
-        unsigned short value;
+        unsigned int bracket_index;
+        std::string address;
```
Comment (MiguelCompany) [inline include/fastdds/rtps/common/Locator.h:295]: I also think this local is never used

```suggestion
```

File: include/fastdds/rtps/common/Locator.h [hunk 8]
```diff
@@ -263,78 +291,78 @@ inline std::istream& operator >>(
...
+        std::string str_locator;
```
Comment (MiguelCompany) [inline include/fastdds/rtps/common/Locator.h:296]: And this one could be declared inside the `try {}` block

**Review by MiguelCompany — APPROVED** [review]

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3256/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/467/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2908/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3288/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3279/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/491/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2930/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3312/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3287/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/499/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2935/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3320/)

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3292/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/503/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2939/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3324/)

**Comment by MiguelCompany** [comment]
Failures seem unrelated.

**Review by MiguelCompany — CHANGES_REQUESTED** [review]

File: src/cpp/utils/IPLocator.cpp [hunk 9]
```diff
@@ -24,14 +24,6 @@ namespace eprosima {
...
 namespace rtps {
 
-IPLocator::IPLocator()
-{
-}
-
-IPLocator::~IPLocator()
-{
-}
-
```
[the remark this answers is not in the document]
Comment (jparisu) [reply src/cpp/utils/IPLocator.cpp:34]: `IPLocator` only consist on `static` functions. Default constructor will be left for API's sake, but will be mark as deprecated.

[the remark this answers is not in the document]
Comment (MiguelCompany) [reply src/cpp/utils/IPLocator.cpp:34]: 🤔 Hmmm. I see. You're right. They can be removed then.

**Review by MiguelCompany — APPROVED** [review]

**Review by MiguelCompany — APPROVED** [review]
LGTM 🏅

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/3319/)
* Linux aarch64[Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux%20aarch64/530/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/2970/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/3351/)

File: src/cpp/utils/IPLocator.cpp [hunk 10]
```diff
@@ -732,6 +779,75 @@ bool IPLocator::isMulticast(
...
+    }
+    return true;
+}
+
+bool IPLocator::IPv6isCorrect(
+        const std::string& ipv6)
+{
+    size_t count = std::count_if( ipv6.begin(), ipv6.end(), []( char c ){return c == ':';});
+
+    // proper number of :
```
[the remark this answers is not in the document]
Comment (RFRIEDM) [reply src/cpp/utils/IPLocator.cpp:868]: Is it intended to follow the RFC? https://tools.ietf.org/html/rfc4291

**Pull request merged**